File: Unit04/activities.py
import array_utils

def print_range (a_range):
    for value in a_range:
        print (value, end=" ")
    print ()

def sum (a_range):
    # Return a sum of the values in the range without using a for loop
    s = 0
    length = len (a_range)
    index = 0
    while index < length:
        value = a_range [index]
        s += value
        index += 1
    return s

def print_odds (an_array):
    length = len (an_array)
    for index in range (length):
        value = an_array [index]
        if value % 2 == 1:
            print (value, end = " ")
    print ()

def print_odds_rec (an_array, index = 0):

    if index == len (an_array):
        return
    else:
        value = an_array [index]
        if value % 2 == 1:
            print (value, end = " ")

    print_odds_rec (an_array, index + 1)

def main ():
    # a = range (10)
    # print_range (a) # Prints the range of 0 to 10 (0, 1, 2...)
    # b = range (5, 15)
    # print_range (b) # Prints the range of 5 to 15 (5, 6, 7...)
    # c = range (20, 80, 4)
    # print_range (c) # Prints every 4th value (20, 24, 28...)
    # a = range (0, 11)
    # print_range (a)
    # b = range (0, 21, 2)
    # print_range (b)
    # c = range (5, 16, 2)
    # print_range (c)
    # d = range (10, -1, -1)
    # print_range(d)
    # a = array_utils.range_array (5, 80, 3)
    # print_odds (a)
    a = array_utils.range_array (6, 80, 3)
    print_odds_rec (a)

if __name__ == "__main__":
    main ()

File: Unit04/Activities.java
public class Activities {
    public static void printOdds (int[] array) {
    for (int index = 0; index < array.length; index++) {
            int value = array[index];
            if (value % 2 != 0) {
                System.out.println(value + " ");
            }
        }
    }

    public static int threeAndFives (int [] array) {
        int sum = 0;
        for (int index = 0; index < array.length; index++) {
            int value = array [index];
            if (value % 3 == 0) { // With both 3 and 5 - if (value % 3 == 0 || value % 5 == 0)
                sum = sum + value;
            }
        }
            return sum;
        }

    public static int printIntegers (int n, int c) {
        // We are trying to only print integers from 0 to N
        // One unit of work is printing a single integer
        // We need to know what integer to print
        // To prepare for the next step we eed to add one to our integer call
        // We know to stop working when our integer call is equal to N
        if (n < 0) {
            return -1;
        }
        if (c == n) {
            System.out.println (c);
            return c;
        } else {
            System.out.println (c); // Unit of work
            int nextC = c + 1; // Prepare for the recursive call
            return c + printIntegers(n, nextC); 
        }
    }

    public static int printIntegers (int n) {
        return printIntegers(n, 0);
    }

    public static int factorial (int n) {
        if (n < 0) {
            return -1;
        } else if (n == 0 || n == 1) {
            return 1;
        } else {
            int result = factorial (n-1);
            int factorial = n * result;
            return factorial;
        }
    }


    public static int gcd(int a, int b) {
        if (a < 1 || b < 1) {
            return -1;
        } else if (a == b){
            return b;
        } else if (a > b) {
            return gcd (a - b, b);
        } else {
            return gcd (b - a, a);
        }
    }
        
public static void main(String[] args) {
    // int [] array = ArrayUtils.rangeArray (50, 75);
    // // printOdds(array);
    // System.out.println(threeAndFives(array));
    // int total = printIntegers(100000);
    // int total = printIntegers(10);
    // System.out.println((total));
    int i = 0;
    while (i < 21) {
        System.out.println("factorial(" + i + ") = " + factorial(i));
        i += 1;
        }
    }
}